### PR TITLE
upgrade object model

### PR DESCRIPTION
**Pull request by panosatha**

### CODE REVIEW

File: flood_adapt/object_model/hazard/measure/floodwall.py
Comment (GundulaW) [inline flood_adapt/object_model/hazard/measure/floodwall.py:25]:
```suggestion
    def load_file(filepath: Union[str, Path]):
```

File: flood_adapt/object_model/hazard/measure/floodwall.py
Comment (GundulaW) [inline flood_adapt/object_model/hazard/measure/floodwall.py:42]: see above

File: flood_adapt/object_model/hazard/physical_projection.py
Comment (GundulaW) [inline flood_adapt/object_model/hazard/physical_projection.py:9]: I think it works like this, but not sure. I did this a while ago but then deleted it in a code revision :-D so you would need to test this
```suggestion
    sea_level_rise: Optional[UnitfulLength]  = UnitfulLength(value=0., units="meters")
```

File: flood_adapt/object_model/hazard/physical_projection.py
Comment (GundulaW) [inline flood_adapt/object_model/hazard/physical_projection.py:10]: same, but I think it's not in the GUI anymore, so we may not need this

File: flood_adapt/object_model/interface/events.py
Comment (GundulaW) [inline flood_adapt/object_model/interface/events.py:9]: same

File: flood_adapt/object_model/interface/events.py
Comment (GundulaW) [inline flood_adapt/object_model/interface/events.py:20]: same

File: flood_adapt/object_model/interface/measures.py
Comment (GundulaW) [inline flood_adapt/object_model/interface/measures.py:20]: same

File: flood_adapt/object_model/interface/projections.py
Comment (GundulaW) [inline flood_adapt/object_model/interface/projections.py:9]: same

File: flood_adapt/object_model/interface/projections.py
Comment (GundulaW) [inline flood_adapt/object_model/interface/projections.py:20]: same

File: flood_adapt/object_model/interface/site.py
Comment (GundulaW) [inline flood_adapt/object_model/interface/site.py:21]: Delete.
The site config file is static data so we don't need a save
```suggestion
```

File: flood_adapt/object_model/interface/site.py
Comment (GundulaW) [inline flood_adapt/object_model/interface/site.py:9]: same

File: flood_adapt/object_model/interface/strategies.py
Comment (GundulaW) [inline flood_adapt/object_model/interface/strategies.py:9]: same

File: flood_adapt/object_model/interface/strategies.py
Comment (GundulaW) [inline flood_adapt/object_model/interface/strategies.py:20]: same

File: flood_adapt/object_model/projection.py
Comment (GundulaW) [inline flood_adapt/object_model/projection.py:38]: same

File: flood_adapt/object_model/projection.py
Comment (GundulaW) [inline flood_adapt/object_model/projection.py:55]: same

File: flood_adapt/object_model/projection.py
Comment (GundulaW) [inline flood_adapt/object_model/projection.py:56]:
```suggestion
        """save Projection to a toml file"""
```

File: flood_adapt/object_model/scenario.py
Comment (GundulaW) [inline flood_adapt/object_model/scenario.py:26]: same

File: flood_adapt/object_model/strategy.py
Comment (GundulaW) [inline flood_adapt/object_model/strategy.py:89]: same

File: tests/test_database/charleston/input/projections/current_empty/current_new.toml
Comment (GundulaW) [inline tests/test_database/charleston/input/projections/current_empty/current_new.toml:4]: Is this the result from a test to write a projection toml? Doesn't seem to work since it is missing all the keys under physical_projection and socio_economic_change. I cannot see the script with teh Projections class in this review but 
I tested writing out an event and that worked. You can have a look there

Comment (GundulaW) [reply tests/test_database/charleston/input/projections/current_empty/current_new.toml:4]: Or is it because default values are not set to zero? We should do that. That will make it easier iin the FIAT and SFINCS adapters

File: flood_adapt/object_model/hazard/hazard.py
Comment (GundulaW) [inline flood_adapt/object_model/hazard/hazard.py:21]: This should be in the scenarios.py script and then imported here. That would be the logical place

File: flood_adapt/object_model/scenario.py
Comment (GundulaW) [inline flood_adapt/object_model/scenario.py:12]: Here would be the logical place for this, remove this from the hazard.py script. I added a comment there as well

```suggestion

class ScenarioModel(BaseModel):
    """BaseModel describing the expected variables and data types of a scenario"""
    name: str
    long_name: str
    event: str
    projection: str
    strategy: str
```

File: flood_adapt/object_model/direct_impacts.py
Comment (GundulaW) [inline flood_adapt/object_model/direct_impacts.py:7]: The ScenarioModel schould be in the scenarios.py script? See respective comments in scenarios and hazard scripts

Comment (GundulaW) [inline flood_adapt/object_model/direct_impacts.py:7]:
```suggestion
from flood_adapt.object_model.hazard.hazard import Hazard
from flood_adapt.object_model.scenario import ScenarioModel
```

File: flood_adapt/object_model/direct_impacts.py
Comment (GundulaW) [inline flood_adapt/object_model/direct_impacts.py:51]: just remove?

**Review by GundulaW — APPROVED** [review]
Puh, lot's of work on your side, big review on my side ;-) 

Some larger comments: 
And is there a reason to use os.PathLike? I used Path and you changed it everywhere. Path from pathlib is what Julian once mentioned to me and Frederique uses to my knowledge as well. Can we use os.pathLike and Path interchangeable? Then it's just a styling question and I do not care too much. I marked them all but you can also just ignore it because as I went on with the review I noticed it's a lot. If Path and os.PathLike are not compatible, we need to make everyone aware of and use it consistently.

Have a look at writing out projections. The newly saved toml looks weird, missing all zero values. We should probably set default values to zero, that will also make our life easier in the SFINCS and FIAT adapter. I did it once but in a version of code a while back. I made a suggestion in the script where you added this as a ToDo but I am not sure if it works. You need to test this. 

The ScenarioModel should be in the scenario script, which would be the logical place where people expect it to be. I made comments/suggested code in all affected scripts.

Minor change but useful: I added the NotImplementedError Julian suggested

I will approve trusting you are making the relevant changes (os.PathLike depending on whether it is compatible with Path, otherwise I leave it up to you if you want to ignore it). I just remember the fuss when I required you to make changes and could not approve it myself anymore

File: flood_adapt/object_model/direct_impacts.py
Comment (panosatha) [reply flood_adapt/object_model/direct_impacts.py:7]: That's exactly what I wanted to do, but I had an issue with cyclic imports when I tried that (basicaly the directimpacts uses the ScenarioModel for typing and it cannot be imported from a module with the class that contains it). I am actually thinking now that we should move all the pydantic class in the interface.

Comment (panosatha) [reply flood_adapt/object_model/direct_impacts.py:7]: see other comment

File: flood_adapt/object_model/hazard/event/synthetic.py
Comment (panosatha) [reply flood_adapt/object_model/hazard/event/synthetic.py:2]: see other comment

File: flood_adapt/object_model/hazard/event/synthetic.py
Comment (panosatha) [reply flood_adapt/object_model/hazard/event/synthetic.py:57]: see other comment

File: flood_adapt/object_model/hazard/event/synthetic.py
Comment (panosatha) [reply flood_adapt/object_model/hazard/event/synthetic.py:74]: see other comment

File: flood_adapt/object_model/hazard/hazard.py
Comment (panosatha) [reply flood_adapt/object_model/hazard/hazard.py:21]: see other comment

File: flood_adapt/object_model/scenario.py
Comment (panosatha) [reply flood_adapt/object_model/scenario.py:12]: see other comment

File: tests/test_database/charleston/input/projections/current_empty/current_new.toml
Comment (panosatha) [reply tests/test_database/charleston/input/projections/current_empty/current_new.toml:4]: Not really. I actually manually made this file to check what happens if for current conditions the toml file is empty and if the defaults zeros are loaded correctly in the object.

**Comment by panosatha** [comment]
- The os.Pathlike is fully compatible with Path. There is a difference to what we use internally as per our agreement (Path from pathlib) and what a user could use for a path variable (could be even a string). I checked online and found this: 
https://stackoverflow.com/questions/58541722/what-is-the-correct-way-in-python-to-annotate-a-path-with-type-hints

- The toml projection you are referring to is not part of the tests, meaning it was not created with the object model. I made it manually to test if an empty toml gets the defaults values.

- I totally aggree with the fact that the ScenarioModel should be in the scenario.py and not hazard.py. That's how I started but I had issues with cyclic imports since we need the ScenarioModel for type hinting in another level (direct_impacts.py). I would actually suggest to move all the pydentic classes in the interface (since either way I think that were they would belong as part of the API) and this way we will not have any ciclic import issues I believe

**Pull request merged**
